pycloudlib/ec2/instance.py

```diff
@@ -81,9 +105,67 @@ def add_network_interface(self) -> str:
         See the AWS documentation for more info:
         https://boto3.readthedocs.io/en/latest/reference/services/ec2.html?#EC2.Client.create_network_interface
         https://boto3.readthedocs.io/en/latest/reference/services/ec2.html?#EC2.Client.attach_network_interface
+
+        Args:
+            ipv4_address_count: number of private ipv4s
+            ipv6_address_count: number of private ipv6s
+            ipv4_public_ip_count: number of public ips associated to private
+                ips.
+            associate_public_ip: created boto3 instance object
+        Returns: private primary ip
         """
         self._log.debug("adding network interface to %s", self.id)
-        interface_id = self._create_network_interface()
+        interface_id = self._create_network_interface(ipv6_address_count)
+        if ipv4_public_ip_count is not None:
+            if (
+                ipv4_address_count is None
+                or ipv4_address_count < ipv4_public_ip_count
+            ):
+                raise PycloudlibError(
+                    f"Invalid configuration:"
+                    f" `ipv4_public_ip_count={ipv4_public_ip_count}` cannot be"
+                    f" greater than `ipv4_address_count={ipv4_address_count}`"
+                )
+
+        if ipv4_address_count is not None and ipv4_address_count > 1:
+            assignment_resp = self._client.assign_private_ip_addresses(
+                NetworkInterfaceId=interface_id,
+                SecondaryPrivateIpAddressCount=ipv4_address_count - 1,
+            )
+            _check_response(assignment_resp)
+            if ipv4_public_ip_count is not None and ipv4_address_count > 1:
+                nics_resp = self._client.describe_network_interfaces(
+                    NetworkInterfaceIds=[interface_id]
+                )
+                _check_response(nics_resp)
+                priv_ips = [
+                    nics_resp["NetworkInterfaces"][0]["PrivateIpAddress"]
+                ]
+                priv_ips.extend(
+                    (
+                        x["PrivateIpAddress"]
+                        for x in assignment_resp["AssignedPrivateIpAddresses"]
+                    )
+                )
+                for i in range(ipv4_address_count):
+                    allo = self._client.allocate_address(Domain="vpc")
```

Comment:
allocate_address allocates an elastic IP address. Shouldn't we only be doing this below in the `if associate_public_ip` scope? This is probably the reason I was seeing limits on public address allocation when running integration tests.

Don't we instead want `assign_private_ip_addresses` here  instead of the allocate_address/associate_address?

Reply:
Yes, iterating over `range(1, ipv4_public_ip_count)` skipping the first NIC as it comes with a public ipv4 by default.

`associate_public_ip` is for a per-NIC global ip, as opposed to adding public ips attached to specific addreses.